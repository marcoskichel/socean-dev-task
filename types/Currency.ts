export enum Currency {
  SOL = 'SOL',
  scnSOL = 'scnSOL',
}
