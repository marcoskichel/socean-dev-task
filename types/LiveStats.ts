export interface LiveStats {
  title: string;
  value: string;
  hint: string;
  progress?: number;
}
