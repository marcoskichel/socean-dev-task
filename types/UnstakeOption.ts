export interface UnstakeOption {
  label: string;
  fee: number;
  value: number;
}
