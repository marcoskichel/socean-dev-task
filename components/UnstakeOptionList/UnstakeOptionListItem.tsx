import { Checkbox, ListItemButton, styled, Typography } from '@mui/material';
import { Currency } from '../../types/Currency';
import { UnstakeOption } from '../../types/UnstakeOption';
import CurrencyIcon from '../CurrencyIcon/CurrencyIcon';

const Wrapper = styled('div')(() => ({
  width: '100%',
  position: 'relative',
}));

const Content = styled('div')(() => ({
  display: 'flex',
  flexDirection: 'column',
  gap: '1rem',
}));

const Description = styled('div')(() => ({
  display: 'flex',
  alignItems: 'center',
}));

const OutcomeOptionCheckbox = styled(Checkbox)(() => ({
  position: 'absolute',
  right: '0',
  top: '1rem',
}));

const Value = styled(Typography)(() => ({
  fontWeight: 600,
  fontSize: '1.5rem',
  marginRight: '1rem',
}));

const StyledListItem = styled(ListItemButton)(() => ({
  '&.Mui-selected': {
    background: 'rgba(242, 79, 131, 0.1)',
    border: '1px solid rgba(242, 79, 131, 0.75)',
  },
}));

type Props = UnstakeOption & {
  id: string;
  selected: boolean;
  onSelect: (value: string) => void;
};

const UnstakeOptionListItem = (props: Props) => {
  const { id, label, fee, value, selected, onSelect } = props;

  const handleSelect = () => {
    onSelect(id);
  };

  return (
    <StyledListItem selected={selected} onClick={handleSelect}>
      <Wrapper>
        <Content>
          <Typography variant="caption">{label}</Typography>
          <Description>
            <CurrencyIcon
              sx={{ padding: 0, marginRight: '0.5rem' }}
              height={26}
              width={26}
              currency={Currency.SOL}
            />
            <Value variant="h5">{`${parseFloat(value.toString())} SOL`}</Value>
            <Typography variant="caption">{`${fee}% fee`}</Typography>
          </Description>
        </Content>
        <OutcomeOptionCheckbox
          color="secondary"
          onSelect={handleSelect}
          checked={selected}
        />
      </Wrapper>
    </StyledListItem>
  );
};

export default UnstakeOptionListItem;
