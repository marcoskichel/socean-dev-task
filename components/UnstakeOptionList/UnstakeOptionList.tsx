import { List } from '@mui/material';
import { UnstakeOption } from '../../types/UnstakeOption';
import UnstakeOptionListItem from './UnstakeOptionListItem';

interface Props {
  selectedId?: string;
  options: UnstakeOption[];
  onChange?: (value: string) => void;
}

/**
 * The list of unstaking options which the user can select from.
 */
const UnstakeOptionList = (props: Props) => {
  const { options, onChange = () => {}, selectedId } = props;

  const handleItemSelect = (optionId: string) => {
    onChange(optionId);
  };

  return (
    <List>
      {options.map(({ id, label, fee, value }) => (
        <UnstakeOptionListItem
          id={id}
          selected={id === selectedId}
          onSelect={handleItemSelect}
          key={id}
          label={label}
          fee={fee}
          value={value}
        />
      ))}
    </List>
  );
};

export default UnstakeOptionList;
