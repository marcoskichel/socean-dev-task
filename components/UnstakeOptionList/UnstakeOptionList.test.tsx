import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UnstakeOption } from '../../types/UnstakeOption';
import UnstakeOptionList from './UnstakeOptionList';

describe('UnstakeOptionList', () => {
  const opts: UnstakeOption[] = [
    { id: 'one', label: 'Test One', value: 1, fee: 0.1 },
    { id: 'two', label: 'Test Two', value: 2, fee: 0.2 },
  ];

  it('renders', () => {
    const { asFragment } = render(
      <UnstakeOptionList options={opts} selectedId={opts[0].id} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the right selected item', async () => {
    const { asFragment } = render(
      <UnstakeOptionList options={opts} selectedId={opts[1].id} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('triggers onChange', async () => {
    const onChange = jest.fn();
    render(
      <UnstakeOptionList
        options={opts}
        selectedId={opts[0].id}
        onChange={onChange}
      />,
    );
    const items = screen.getAllByRole('button');
    userEvent.click(items[1]);
    expect(onChange).toHaveBeenLastCalledWith(opts[1].id);
  });
});
