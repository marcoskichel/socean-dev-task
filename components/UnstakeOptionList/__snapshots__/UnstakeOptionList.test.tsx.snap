// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnstakeOptionList displays the right selected item 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-zvqzn8-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="css-1kxonj9"
      >
        <div
          class="css-19s3t1h"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
          >
            Test One
          </span>
          <div
            class="css-70qvj9"
          >
            <div
              class="css-cleg7o"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
              >
                <img
                  alt="Currency"
                  data-nimg="fixed"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1kt30nm-MuiTypography-root"
            >
              1 SOL
            </h5>
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              0.1% fee
            </span>
          </div>
        </div>
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1p28g3s-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiButtonBase-root css-zvqzn8-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="css-1kxonj9"
      >
        <div
          class="css-19s3t1h"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
          >
            Test Two
          </span>
          <div
            class="css-70qvj9"
          >
            <div
              class="css-cleg7o"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
              >
                <img
                  alt="Currency"
                  data-nimg="fixed"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1kt30nm-MuiTypography-root"
            >
              2 SOL
            </h5>
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              0.2% fee
            </span>
          </div>
        </div>
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-1p28g3s-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </ul>
</DocumentFragment>
`;

exports[`UnstakeOptionList renders 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      class="MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiButtonBase-root css-zvqzn8-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="css-1kxonj9"
      >
        <div
          class="css-19s3t1h"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
          >
            Test One
          </span>
          <div
            class="css-70qvj9"
          >
            <div
              class="css-cleg7o"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
              >
                <img
                  alt="Currency"
                  data-nimg="fixed"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1kt30nm-MuiTypography-root"
            >
              1 SOL
            </h5>
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              0.1% fee
            </span>
          </div>
        </div>
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-1p28g3s-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-zvqzn8-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="css-1kxonj9"
      >
        <div
          class="css-19s3t1h"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
          >
            Test Two
          </span>
          <div
            class="css-70qvj9"
          >
            <div
              class="css-cleg7o"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
              >
                <img
                  alt="Currency"
                  data-nimg="fixed"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1kt30nm-MuiTypography-root"
            >
              2 SOL
            </h5>
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              0.2% fee
            </span>
          </div>
        </div>
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1p28g3s-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </ul>
</DocumentFragment>
`;
